# OTT Customer Churn Analytics

## Data Cleaning and Preprocessing

**Author:** Omar Rashed Sizan  
**Role:** Data Analyst Portfolio Project  
**Tools:** MySQL, Python (Pandas, NumPy), Jupyter Notebook  
**Database:** customer_churn  

---

### Notebook Objective

This notebook performs data cleaning and preprocessing on the raw operational database to create a reliable, standardized, and analysis-ready dataset for customer churn analytics.

# 1. Project Objective

## Objective

The primary objective of this notebook is to perform a comprehensive data cleaning and preprocessing process for the **OTT Customer Churn Analytics** project.

The raw data extracted from the operational MySQL database contains various real-world data quality issues, including missing values, inconsistent categorical values, invalid business records, and rule violations. These issues can negatively impact analytical accuracy, business reporting, dashboard development, and predictive modeling.

The purpose of this notebook is to identify, validate, and resolve these issues using Python while preserving the business meaning of the data.

The cleaned dataset produced through this process will serve as the single source of truth for all subsequent stages of the project, including Exploratory Data Analysis (EDA), Business Intelligence dashboards, and churn analytics.

---

## Cleaning Objectives

This notebook aims to achieve the following objectives:

- Connect directly to the MySQL database and import all required tables.
- Validate data types and overall dataset structure.
- Identify and handle missing values.
- Standardize inconsistent categorical values.
- Correct invalid records based on predefined business rules.
- Verify referential integrity across related tables.
- Perform feature engineering to improve analytical capability.
- Produce a clean, reliable, and analysis-ready dataset.
- Export the cleaned data for downstream analytics.

---

## Expected Outcome

Upon completion of this notebook, the dataset will:

- Maintain consistent and standardized values.
- Satisfy all critical business validation rules.
- Preserve relational integrity between tables.
- Be suitable for statistical analysis and visualization.
- Support accurate KPI reporting.
- Provide a reliable foundation for customer churn analysis and business decision-making.

---

## Business Value

High-quality data is a prerequisite for generating trustworthy business insights.

By improving data quality before analysis, this notebook helps ensure that business decisions related to customer retention, revenue optimization, subscription strategy, and customer support are based on accurate and reliable information rather than inconsistent or erroneous data.

This preprocessing stage establishes the foundation for the remaining phases of the project, including Exploratory Data Analysis (EDA), dashboard development, and business recommendations.

# 2. Business Context

## Business Background

The company is an OTT (Over-the-Top) video streaming platform that offers subscription-based digital entertainment services to customers across multiple countries.

In recent months, the management team has observed an increase in customer churn, leading to declining recurring revenue, lower customer lifetime value (CLTV), and higher customer acquisition costs (CAC). Since acquiring a new customer is significantly more expensive than retaining an existing one, improving customer retention has become a strategic business priority.

To address this challenge, the executive team initiated a customer churn analytics project with the objective of identifying the key factors that influence customer cancellations and providing data-driven recommendations to improve retention.

---

## Business Problem

The business aims to answer several critical questions, including:

- Which customer segments are most likely to churn?
- Does subscription type influence customer retention?
- How do customer support experiences impact churn?
- Which complaint categories are associated with low customer satisfaction?
- Are Monthly subscribers more likely to churn than Annual subscribers?
- Which acquisition channels generate the most loyal customers?
- Which customer groups should be prioritized for retention initiatives?

Answering these questions requires a reliable and high-quality dataset.

---

## Why Data Cleaning is Required

The raw operational database was collected from multiple business processes, including customer registration, subscription management, and customer support operations.

During the SQL data profiling phase, several data quality issues were identified, including:

- Missing values
- Inconsistent categorical values
- Invalid business values
- Business rule violations
- Data standardization issues

If these issues are not resolved before analysis, they can lead to inaccurate KPIs, misleading dashboards, biased analytical results, and incorrect business decisions.

Therefore, a structured data cleaning process is essential before performing Exploratory Data Analysis (EDA) and business reporting.

---

## Scope of This Notebook

This notebook focuses exclusively on preparing the data for analysis.

The preprocessing activities include:

- Connecting to the MySQL database
- Loading all project tables
- Assessing data quality
- Cleaning and standardizing data
- Validating business rules
- Creating analysis-ready datasets

The cleaned datasets generated from this notebook will be used in the next phases of the project, including Exploratory Data Analysis (EDA), Business Intelligence dashboard development, and business recommendation generation.

---

## Expected Business Outcome

By completing this preprocessing stage, the organization will obtain a reliable, consistent, and analysis-ready dataset that supports accurate customer churn analysis.

The cleaned data will enable stakeholders to:

- Measure churn accurately.
- Understand customer behavior.
- Identify high-risk customer segments.
- Improve retention strategies.
- Support data-driven decision-making.
- Build reliable dashboards and analytical reports.

Ultimately, this notebook lays the foundation for transforming raw operational data into actionable business insights that can help reduce customer churn and improve long-term business performance.

# 3. Import Required Libraries

## Overview

This section imports all the Python libraries required for data loading, exploration, cleaning, preprocessing, and database connectivity.

Each library has a specific role in the data preparation workflow. Using only the necessary libraries helps keep the notebook organized, efficient, and easy to maintain.

---

## Libraries Used

| Library | Purpose |
|---------|---------|
| pandas | Data manipulation and analysis |
| numpy | Numerical operations and handling missing values |
| sqlalchemy | Connect Python with MySQL database |
| pymysql | MySQL database driver for SQLAlchemy |
| warnings | Suppress unnecessary warning messages |

---

## Why These Libraries?

- **Pandas** provides powerful DataFrame structures for cleaning, transforming, and analyzing tabular data.
- **NumPy** supports efficient numerical computations and missing value handling.
- **SQLAlchemy** offers a robust and industry-standard interface for connecting Python applications to relational databases.
- **PyMySQL** acts as the MySQL driver required by SQLAlchemy.
- **Warnings** is used to suppress non-critical warning messages, ensuring cleaner notebook outputs.

In [1]:
pip install ipykernel

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 26.1.2 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [2]:
 pip install --upgrade pip

   ---------------------------------------- 0.0/1.8 MB ? eta -:--:--
   ----- ---------------------------------- 0.3/1.8 MB ? eta -:--:--
   ----------------------- ---------------- 1.0/1.8 MB 3.0 MB/s eta 0:00:01
   ---------------------------- ----------- 1.3/1.8 MB 2.1 MB/s eta 0:00:01
   ---------------------------------------- 1.8/1.8 MB 2.2 MB/s  0:00:00
  Attempting uninstall: pip
    Found existing installation: pip 26.1.2
    Uninstalling pip-26.1.2:
      Successfully uninstalled pip-26.1.2
Note: you may need to restart the kernel to use updated packages.


In [3]:
pip install sqlalchemy

Note: you may need to restart the kernel to use updated packages.


In [4]:
pip install pymysql

In [5]:
# =====================================================
# Data Manipulation
# =====================================================

import pandas as pd
import numpy as np

# =====================================================
# Database Connection
# =====================================================

from sqlalchemy import create_engine
import pymysql

# =====================================================
# Notebook Settings
# =====================================================

import warnings
warnings.filterwarnings("ignore")

# =====================================================
# Display Settings
# =====================================================

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)
pd.set_option("display.width", None)
pd.set_option("display.max_colwidth", 100)

print("Libraries imported successfully.")

Libraries imported successfully.


# 4. Connect to MySQL Database

## Overview

This project uses a MySQL database (`customer_churn`) as the primary data source.

Instead of importing CSV files manually, a direct database connection is established using SQLAlchemy and PyMySQL. This approach reflects real-world data analytics workflows, where analysts retrieve data directly from relational databases.

The connection engine created in this section will be reused throughout the notebook to query and load data efficiently.

In [6]:
from sqlalchemy import create_engine

# Database Credentials
USERNAME = "root"
PASSWORD = "4march2023"
HOST = "localhost"
PORT = "3306"
DATABASE = "customer_churn"

# Create Database Engine
engine = create_engine(
    f"mysql+pymysql://{USERNAME}:{PASSWORD}@{HOST}:{PORT}/{DATABASE}"
)

print("Database Engine Created Successfully.")

Database Engine Created Successfully.


In [7]:
try:
    with engine.connect() as conn:
        print("="*60)
        print(" Connected to MySQL Successfully")
        print(f" Database : {DATABASE}")
        print("="*60)

except Exception as e:
    print("Connection Failed!")
    print(e)

 Connected to MySQL Successfully
 Database : customer_churn


# Load Table

In [8]:
# Load Tables

customer_df = pd.read_sql("SELECT * FROM db_customer", engine)

subscription_df = pd.read_sql("SELECT * FROM db_subscription", engine)

support_df = pd.read_sql("SELECT * FROM db_support", engine)

In [9]:
customer_df.head()

,customerid,name,country,state,gender,dob,interests,pincode
0,CUST-0001,Bikash Chowdhury,Bangladesh,Chittagong,Female,1966-06-05,Comedy,1434
1,CUST-0002,Pooja Ahmed,india,Tamil Nadu,male,1980-04-06,"Drama, Thriller",2519
2,CUST-0003,David Chowdhury,Bangladesh,Dhakaa,None,1978-02-25,"Sports, Fantasy, Romance",1750
3,CUST-0004,Prakash Alam,India,Delhi,F,1980-12-08,Kids,2169
4,CUST-0005,Rohan Gupta,Bangladesh,Rajshahi,Female,1979-07-20,Thriller,7572


In [10]:
 subscription_df.head()

,customerid,subscription_start_date,subscription_type,renewal_date,plan_type,contract_type,cancellation_date,cancellation_reason,monthly_charges,cltv,churn_score
0,CUST-0001,2019-01-15,Referral,2020-01-15,Standard,Annual,None,None,12.99,820,18
1,CUST-0002,2019-02-03,Organic,2019-03-05,Basic,Monthly,2020-06-10,No Longer Interested,7.49,210,67
2,CUST-0003,2019-02-27,Paid,2020-02-27,Premium,Annual,None,None,19.99,1450,22
3,CUST-0004,2019-03-12,REFERRAL,2020-03-10,Standard,Annual,2021-04-18,Too Expensive,13.49,560,54
4,CUST-0005,2019-03-28,Organic,2019-04-28,Basic,Monthly,2019-07-02,Forgot to Cancel Trial,6.99,95,71


In [11]:
support_df.head()

,support_id,customerid,complaint_date,escalations,csat_score,col_1,comment
0,1,CUST-0042,2020-02-14,0,4.0,Streaming,Video buffers every evening
1,2,CUST-0120,2021-11-03,1,3.0,Billing,Payment deducted twice
2,3,CUST-0120,2022-01-15,1,2.0,Billing,Still waiting for refund
3,4,CUST-0333,2023-06-20,0,5.0,Login,Cannot login after password reset
4,5,CUST-0333,2023-06-20,0,5.0,Login,Cannot login after password reset


In [12]:
print("="*60)

print("Customer :", customer_df.shape)

print("Subscription :", subscription_df.shape)

print("Support :", support_df.shape)

print("="*60)

Customer : (600, 8)
Subscription : (600, 11)
Support : (648, 7)


### Observation

The dataset was successfully loaded from the MySQL database into three separate Pandas DataFrames without any loading errors.

- **db_customer** contains **600 records** and **8 columns**, representing the master customer information. Each record corresponds to a unique customer.

- **db_subscription** contains **600 records** and **11 columns**, indicating that every customer has one subscription record. This suggests a **one-to-one relationship** between the customer and subscription tables.

- **db_support** contains **648 records** and **7 columns**, which is greater than the number of customers. This indicates that some customers have contacted customer support multiple times, representing a **one-to-many relationship** between customers and support tickets.

From a business perspective, this is expected in an OTT streaming platform, where customers may raise multiple complaints during their subscription lifecycle. These repeated interactions are valuable for identifying customer dissatisfaction, support burden, and potential churn behavior.

The successful loading of all three datasets confirms that the data is ready for the initial inspection and quality assessment phase.

In [13]:
customer_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 600 entries, 0 to 599
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   customerid  600 non-null    object        
 1   name        600 non-null    object        
 2   country     555 non-null    object        
 3   state       559 non-null    object        
 4   gender      570 non-null    object        
 5   dob         600 non-null    datetime64[ns]
 6   interests   561 non-null    object        
 7   pincode     583 non-null    object        
dtypes: datetime64[ns](1), object(7)
memory usage: 37.6+ KB


### Observation

The `db_customer` table contains **600 records** and **8 columns**, with a total memory usage of approximately **37.6 KB**. The dataset primarily consists of categorical (`object`) variables and one date column (`datetime64[ns]`), indicating that it stores customer demographic and profile information.

#### Key Findings

- The `customerid` column is populated for all records, ensuring that each customer can be uniquely identified.
- The `dob` column has been correctly imported as a **datetime** data type, eliminating the need for additional date conversion during preprocessing.
- The remaining columns are stored as **object** data types, which is appropriate because they represent categorical information such as customer demographics and preferences.
- The dataset structure is suitable for customer segmentation, demographic analysis, and customer-level feature engineering.

#### Data Quality Observation

The output indicates that some columns have fewer non-null values than the total number of records. This suggests the presence of missing values in a few customer attributes. A detailed missing value assessment will be performed in the dedicated **Data Quality Assessment** section later in this notebook.

#### Business Perspective

The `db_customer` table serves as the **master customer dimension table** within the OTT platform. It contains the core demographic and profile information that will later be integrated with subscription and support data to analyze customer behavior and identify potential churn patterns.

In [14]:
subscription_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 600 entries, 0 to 599
Data columns (total 11 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   customerid               600 non-null    object 
 1   subscription_start_date  600 non-null    object 
 2   subscription_type        599 non-null    object 
 3   renewal_date             599 non-null    object 
 4   plan_type                600 non-null    object 
 5   contract_type            600 non-null    object 
 6   cancellation_date        329 non-null    object 
 7   cancellation_reason      330 non-null    object 
 8   monthly_charges          600 non-null    float64
 9   cltv                     600 non-null    int64  
 10  churn_score              600 non-null    int64  
dtypes: float64(1), int64(2), object(8)
memory usage: 51.7+ KB


### Observation

The `db_subscription` table contains **600 records** and **11 columns**, with an approximate memory usage of **51.7 KB**. This table stores customer subscription details, including subscription lifecycle, pricing, and churn-related information.

#### Key Findings

- The `customerid` column is fully populated, ensuring that every subscription record can be linked to a corresponding customer.
- The numerical columns (`monthly_charges`, `cltv`, and `churn_score`) have been imported using appropriate numeric data types (`float64` and `int64`), making them ready for statistical analysis and feature engineering.
- Most remaining columns are stored as `object` data types because they represent categorical or date-related information.

#### Data Quality Observation

The dataset contains several date-related fields, including:

- `subscription_start_date`
- `renewal_date`
- `cancellation_date`

However, these columns have been imported as **object** instead of **datetime**. Before performing any time-based analysis, these fields should be converted into the proper `datetime` format during the data cleaning phase.

The output also indicates that a few columns contain fewer non-null values than the total number of records. This suggests the presence of missing values, particularly in subscription-related attributes. A detailed completeness assessment will be conducted later in the dedicated **Data Quality Assessment** section.

It is also important to note that `cancellation_date` contains a large number of missing values. From a business perspective, this is expected because active subscribers who have not cancelled their subscriptions will naturally have no cancellation date recorded.

#### Business Perspective

The `db_subscription` table is the **core fact table** of this project. It captures each customer's subscription lifecycle, pricing information, contract details, and churn indicators. Most business analyses—including customer retention, churn prediction, customer lifetime value (CLTV), and revenue analysis—will primarily rely on this table. Therefore, ensuring the quality and consistency of this dataset is critical before proceeding with exploratory analysis and predictive modeling.

In [15]:
support_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 648 entries, 0 to 647
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   support_id      648 non-null    int64         
 1   customerid      648 non-null    object        
 2   complaint_date  648 non-null    datetime64[ns]
 3   escalations     648 non-null    object        
 4   csat_score      640 non-null    float64       
 5   col_1           648 non-null    object        
 6   comment         616 non-null    object        
dtypes: datetime64[ns](1), float64(1), int64(1), object(4)
memory usage: 35.6+ KB


### Observation

The `db_support` table contains **648 records** and **7 columns**, with an approximate memory usage of **35.6 KB**. This table stores customer support interactions, including complaint history, escalation status, customer satisfaction scores, issue categories, and customer feedback.

#### Key Findings

- The `support_id` column contains **648 non-null values** and serves as the unique identifier for each support ticket.
- The `customerid` column is fully populated, allowing every support ticket to be linked with a customer record.
- The `complaint_date` column has been correctly imported as the **datetime** data type, making it suitable for time-based analysis such as complaint trends and support response analysis.
- The `csat_score` column has been imported as a numeric (`float64`) variable, allowing customer satisfaction metrics to be calculated directly.
- Most remaining fields are categorical (`object`) variables representing issue type, escalation status, and customer comments.

#### Data Quality Observation

The output indicates that a few columns contain fewer non-null values than the total number of records, suggesting the presence of missing values. These will be analyzed in detail during the **Data Quality Assessment** section.

In addition, the `escalations` column has been imported as an **object** data type. Based on the business definition, this field is expected to represent categorical escalation levels (e.g., 0 = First-Level Support, 1 = Escalated). Therefore, its values will be validated and converted into the appropriate data type during the data cleaning process.

#### Business Perspective

The `db_support` table captures customer interactions after service usage and provides valuable operational insights into customer experience. Support history, escalation frequency, issue categories, and customer satisfaction (CSAT) are strong behavioral indicators that can help identify dissatisfied customers and improve churn prediction models.

When combined with customer demographics and subscription information, this dataset enables a comprehensive analysis of customer behavior, service quality, and retention risk.

In [16]:
support_df.rename(
    columns={"col_1": "issue_category"},
    inplace=True
)

In [17]:
customer_df.isnull().sum()*100 / len(customer_df)

customerid    0.000000
name          0.000000
country       7.500000
state         6.833333
gender        5.000000
dob           0.000000
interests     6.500000
pincode       2.833333
dtype: float64

### Observation

The missing value analysis shows that the `db_customer` table has a relatively high level of completeness, with only a few missing values in selected demographic attributes.

| Column | Missing Values | Percentage | Assessment |
|--------|---------------:|-----------:|------------|
| customerid | 0 | 0.00% | Complete |
| name | 0 | 0.00% | Complete |
| country | 45 | 7.50% | Moderate Missing |
| state | 41 | 6.83% | Moderate Missing |
| gender | 30 | 5.00% | Low Missing |
| dob | 0 | 0.00% | Complete |
| interests | 39 | 6.50% | Moderate Missing |
| pincode | 17 | 2.83% | Low Missing |

#### Key Findings

- The primary identifier (`customerid`) contains **no missing values**, ensuring reliable joins across all tables.
- Critical customer attributes such as **customer name** and **date of birth** are fully populated.
- Missing values are mainly concentrated in demographic and preference-related fields, including **country**, **state**, **gender**, **interests**, and **pincode**.
- No single column has more than **10% missing values**, indicating that the overall completeness of the dataset is good.

#### Business Perspective

Although these missing values do not prevent churn analysis, they can reduce the quality of customer segmentation and demographic reporting. Missing values in geographic fields (`country`, `state`, `pincode`) may affect regional performance analysis, while missing values in `interests` can reduce the effectiveness of customer preference analysis and recommendation-related insights.

#### Cleaning Strategy

Since the percentage of missing values is relatively low, records will not be removed. Instead, appropriate imputation techniques or placeholder values (e.g., `"Unknown"`) will be considered during the data cleaning phase to preserve as much information as possible without introducing significant bias.

In [18]:
subscription_df.isnull().sum()*100 / len(subscription_df)

customerid                  0.000000
subscription_start_date     0.000000
subscription_type           0.166667
renewal_date                0.166667
plan_type                   0.000000
contract_type               0.000000
cancellation_date          45.166667
cancellation_reason        45.000000
monthly_charges             0.000000
cltv                        0.000000
churn_score                 0.000000
dtype: float64

### Observation

The missing value assessment of the `db_subscription` table indicates that most business-critical columns are complete. Only a few fields contain missing values, and not all of them represent data quality issues.

| Column | Missing (%) | Assessment |
|--------|------------:|------------|
| customerid | 0.00% | Complete |
| subscription_start_date | 0.00% | Complete |
| subscription_type | 0.17% | Very Low Missing |
| renewal_date | 0.17% | Very Low Missing |
| plan_type | 0.00% | Complete |
| contract_type | 0.00% | Complete |
| cancellation_date | 45.17% | Expected Missing |
| cancellation_reason | 45.00% | Expected Missing |
| monthly_charges | 0.00% | Complete |
| cltv | 0.00% | Complete |
| churn_score | 0.00% | Complete |

#### Key Findings

- Most columns are fully populated, indicating a high level of data completeness.
- Only **one record (0.17%)** is missing `subscription_type` and `renewal_date`, which is considered negligible and can be handled during data cleaning.
- Nearly **45% of records** have missing values in `cancellation_date` and `cancellation_reason`.

#### Business Perspective

The missing values in `cancellation_date` and `cancellation_reason` are **business-expected** rather than data quality issues. Customers who are still actively subscribed have not cancelled their subscriptions; therefore, these fields naturally remain empty.

Consequently, these missing values should **not** be imputed with arbitrary values or treated as errors. Instead, they represent active customers and should be preserved during data preprocessing.

#### Cleaning Strategy

- Investigate the single missing values in `subscription_type` and `renewal_date` to determine the most appropriate treatment (imputation or removal).
- Retain the missing values in `cancellation_date` and `cancellation_reason`, as they carry meaningful business information indicating active subscriptions.

In [19]:
support_df.isnull().sum()*100 / len(support_df)

support_id        0.000000
customerid        0.000000
complaint_date    0.000000
escalations       0.000000
csat_score        1.234568
issue_category    0.000000
comment           4.938272
dtype: float64

### Observation

The `db_support` table demonstrates a high level of completeness, with only two columns containing missing values. Most operational fields required for customer support analysis are fully populated.

| Column | Missing (%) | Assessment |
|--------|------------:|------------|
| support_id | 0.00% | Complete |
| customerid | 0.00% | Complete |
| complaint_date | 0.00% | Complete |
| escalations | 0.00% | Complete |
| csat_score | 1.23% | Low Missing |
| issue_category | 0.00% | Complete |
| comment | 4.94% | Moderate Missing |

#### Key Findings

- All primary operational fields (`support_id`, `customerid`, `complaint_date`, `escalations`, and `issue_category`) are fully populated.
- Only **1.23%** of support records are missing a `csat_score`, indicating that a small number of customers did not provide satisfaction feedback after their support interaction.
- Approximately **4.94%** of records have missing values in the `comment` column.

#### Business Perspective

The missing values in `csat_score` most likely indicate that some customers did not participate in the post-support satisfaction survey. This is a common scenario in customer support systems and does not necessarily represent poor data quality.

Similarly, missing values in the `comment` field are generally expected. Not every support interaction requires a written customer comment, especially when issues are resolved quickly or customers choose not to leave additional feedback.

Therefore, these missing values should be evaluated carefully during preprocessing rather than automatically treated as errors.

#### Cleaning Strategy

- Investigate the missing `csat_score` records and determine whether they should remain as missing or be excluded from analyses requiring customer satisfaction metrics.
- Preserve missing values in the `comment` column, as the absence of customer feedback is itself meaningful information and should not be replaced with artificial text.

# Duplicate

In [20]:
# Check duplicate rows
customer_df.duplicated().sum()

0

### Observation

The duplicate analysis confirms that the **`db_customer`** table contains **no duplicate records**.

#### Key Findings

- Total Duplicate Records: **0**
- Each customer profile is stored only once in the dataset.
- The table maintains a one-record-per-customer structure, which is essential for accurate customer-level analysis.

#### Business Perspective

Duplicate customer records can lead to inaccurate customer counts, incorrect segmentation, and biased analytical results. The absence of duplicate records indicates good data quality and ensures reliable downstream analysis.

In [21]:
subscription_df.duplicated().sum()

0

### Observation

The duplicate analysis shows that the **`db_subscription`** table contains **no duplicate records**.

#### Key Findings

- Total Duplicate Records: **0**
- Each subscription entry represents a unique customer subscription.

#### Business Perspective

Duplicate subscription records could artificially inflate revenue, customer counts, and churn metrics. The absence of duplicate records confirms that the subscription data is structurally reliable for business analysis.

In [22]:
support_df.duplicated().sum()

0

### Observation

The **`db_support`** table contains **no fully duplicated records**.

#### Key Findings

- Total Duplicate Records: **0**
- Every support ticket represents a unique support interaction.

#### Business Perspective

Although customers may submit multiple complaints over time, each support record represents a separate business event. Therefore, the absence of duplicate rows confirms that support interactions have been recorded correctly without accidental duplication.

# Categorical Value Analysis

Number of Unique Values

In [23]:
customer_df["country"].nunique(dropna=False)

10

Value Distribution

In [24]:
customer_df["country"].value_counts(dropna=False)

country
Bangladesh    184
India          80
BD             70
bangladesh     55
None           45
IND            41
india          35
Nepal          31
NP             31
nepal          28
Name: count, dtype: int64

Missing Values

In [25]:
customer_df["country"].isnull().sum()

45

### Observation

The `country` column contains **10 distinct values (including missing values)**. However, the analysis reveals several inconsistencies in how country names have been recorded.

#### Value Distribution

| Country Value | Count |
|---------------|------:|
| Bangladesh | 184 |
| India | 80 |
| BD | 70 |
| bangladesh | 55 |
| Missing | 45 |
| IND | 41 |
| india | 35 |
| Nepal | 31 |
| NP | 31 |
| nepal | 28 |

#### Data Quality Issues Identified

The analysis identified three major data quality issues:

1. **Case Inconsistency**
   - `Bangladesh` vs `bangladesh`
   - `India` vs `india`
   - `Nepal` vs `nepal`

2. **Abbreviation Inconsistency**
   - `Bangladesh` vs `BD`
   - `India` vs `IND`
   - `Nepal` vs `NP`

3. **Missing Values**
   - The column contains **45 missing values (7.5%)**, which may affect geographical reporting and customer segmentation.

#### Business Impact

If these inconsistencies are not corrected, the same country will be treated as multiple categories during analysis. For example, **Bangladesh**, **bangladesh**, and **BD** would appear as three separate countries in reports and dashboards, resulting in inaccurate customer distribution and misleading business insights.

#### Cleaning Strategy

During the preprocessing phase:

- Country names will be standardized into a single naming convention.
- Abbreviations (e.g., `BD`, `IND`, `NP`) will be converted to their full country names.
- Letter casing will be standardized using title case.
- Missing values will be handled according to the project's predefined data cleaning strategy.

In [26]:
customer_df["state"].nunique(dropna=False)

19

In [27]:
customer_df["state"].value_counts(dropna=False)

state
Khulna         70
Sylhet         67
Rajshahi       57
None           41
West Bengal    34
Maharashtra    33
Tamil Nadu     33
Gandaki        32
Delhi          30
Karnataka      29
Chittagong     28
Chattogram     28
Lumbini        27
Bagmati        20
DHAKA          18
Dhaka          17
Dhakaa         15
dhaka          14
Kathmandu       7
Name: count, dtype: int64

In [28]:
customer_df["state"].isnull().sum()

41

### Observation

The `state` column contains **19 unique values (including missing values)** representing customer locations across Bangladesh, India, and Nepal. While most state names are valid, the analysis identified several inconsistencies that should be standardized before analysis.

#### Value Distribution

| State Value | Count |
|-------------|------:|
| Khulna | 70 |
| Sylhet | 67 |
| Rajshahi | 57 |
| Missing | 41 |
| West Bengal | 34 |
| Maharashtra | 33 |
| Tamil Nadu | 33 |
| Gandaki | 32 |
| Delhi | 30 |
| Karnataka | 29 |
| Chittagong | 28 |
| Chattogram | 28 |
| Lumbini | 27 |
| Bagmati | 20 |
| DHAKA | 18 |
| Dhaka | 17 |
| Dhakaa | 15 |
| dhaka | 14 |
| Kathmandu | 7 |

#### Data Quality Issues Identified

The analysis identified three major data quality issues:

1. **Case Inconsistency**
   - `Dhaka`
   - `DHAKA`
   - `dhaka`

2. **Spelling Inconsistency**
   - `Dhaka`
   - `Dhakaa`

3. **Alternative Naming**
   - `Chittagong`
   - `Chattogram`

4. **Missing Values**
   - The column contains **41 missing values (6.83%)**, which may affect location-based analysis and regional reporting.

#### Business Impact

Inconsistent state names can fragment customer distribution across multiple categories. For example, `Dhaka`, `DHAKA`, `dhaka`, and `Dhakaa` all represent the same location but would be treated as four different regions during analysis. Likewise, `Chittagong` and `Chattogram` refer to the same administrative division but would appear separately in reports.

Such inconsistencies can lead to inaccurate regional dashboards, customer segmentation, and geographic trend analysis.

#### Cleaning Strategy

During preprocessing:

- All state names will be converted to a consistent title case format.
- `Dhakaa` will be standardized to `Dhaka`.
- `DHAKA` and `dhaka` will be converted to `Dhaka`.
- `Chittagong` and `Chattogram` will be standardized using a single business-approved naming convention.
- Missing values will be handled according to the project's predefined missing value strategy.

In [29]:
customer_df["gender"].nunique(dropna=False)

10

In [30]:
customer_df["gender"].value_counts(dropna=False)

gender
Female     104
Male        91
male        90
female      90
Men         48
F           44
M           41
Women       39
None        30
Unknown     23
Name: count, dtype: int64

In [31]:
customer_df["gender"].isnull().sum()

30

### Observation

The `gender` column contains **10 unique values (including missing values)**. The analysis reveals multiple representations of the same gender category, resulting in inconsistent categorical data.

#### Value Distribution

| Gender Value | Count |
|--------------|------:|
| Female | 104 |
| Male | 91 |
| male | 90 |
| female | 90 |
| Men | 48 |
| F | 44 |
| M | 41 |
| Women | 39 |
| Missing | 30 |
| Unknown | 23 |

#### Data Quality Issues Identified

The analysis identified the following data quality issues:

1. **Case Inconsistency**
   - `Male` vs `male`
   - `Female` vs `female`

2. **Abbreviation Inconsistency**
   - `M` vs `Male`
   - `F` vs `Female`

3. **Synonym Inconsistency**
   - `Men` represents `Male`
   - `Women` represents `Female`

4. **Unknown Category**
   - `Unknown` appears as a separate category, indicating that gender information was unavailable for some customers.

5. **Missing Values**
   - The column contains **30 missing values (5.00%)**.

#### Business Impact

Inconsistent gender values can significantly affect demographic analysis, customer segmentation, and marketing reports. For example, `Male`, `male`, `M`, and `Men` all represent the same customer group but would be counted as four separate categories if not standardized.

Similarly, `Female`, `female`, `F`, and `Women` represent the same category and should be consolidated before performing any demographic analysis or visualization.

#### Cleaning Strategy

During preprocessing:

- Convert all gender values to a standardized format.
- Standardize:
  - `male`, `M`, `Men` → `Male`
  - `female`, `F`, `Women` → `Female`
- Preserve `Unknown` as a valid business category unless business requirements specify otherwise.
- Handle missing values according to the project's missing value strategy.

In [32]:
customer_df["interests"].nunique(dropna=False)

345

In [33]:
customer_df["interests"].value_counts(dropna=False)

interests
None                        39
Fantasy                     18
Drama                       18
Reality TV                  14
Action                      13
                            ..
Fantasy, Romance, Action     1
Kids, Comedy, Thriller       1
Action, Sci-Fi               1
Anime, Romance               1
Comedy, Sports               1
Name: count, Length: 345, dtype: int64

In [34]:
customer_df["interests"].isnull().sum()

39

### Observation

The `interests` column contains **345 unique values (including missing values)**, indicating a high-cardinality categorical feature. Unlike other categorical columns, many records contain multiple interests separated by commas, representing customers' content preferences.

#### Key Findings

- Total Unique Values: **345**
- Missing Values: **39 (6.50%)**
- Most records contain one or more entertainment preferences such as:
  - Action
  - Drama
  - Fantasy
  - Reality TV
  - Comedy
  - Sports
  - Romance
  - Anime
  - Sci-Fi

Additionally, many customers have multiple interests recorded in a single field, for example:

- Fantasy, Romance, Action
- Kids, Comedy, Thriller
- Action, Sci-Fi
- Anime, Romance
- Comedy, Sports

#### Data Quality Assessment

The analysis identified the following observations:

1. **High Cardinality**
   - The column contains 345 unique combinations because multiple interests are stored together in a single field.

2. **Multi-Value Attributes**
   - Customer preferences are stored as comma-separated values rather than separate normalized records.

3. **Missing Values**
   - The column contains **39 missing values (6.50%)**.

#### Business Impact

The `interests` column provides valuable information about customer viewing preferences and can be used for customer segmentation, recommendation systems, and churn prediction.

Although the column has many unique values, this is expected because customers may select multiple content categories. Therefore, the high number of unique values does **not** indicate poor data quality.

However, storing multiple interests in a single column limits analytical flexibility. For advanced analytics or machine learning, these values may need to be split into individual categories or transformed using techniques such as one-hot encoding or multi-label encoding.

#### Cleaning Strategy

During preprocessing:

- Remove unnecessary leading and trailing spaces.
- Standardize text formatting (e.g., consistent title case if required).
- Preserve comma-separated values, as they represent valid business information.
- Handle missing values according to the project's predefined missing value strategy.
- For machine learning, the column may later be transformed into multiple binary features using multi-label encoding.

In [35]:
subscription_df["subscription_type"].nunique(dropna=False)

7

In [36]:
subscription_df["subscription_type"].value_counts(dropna=False)

subscription_type
Organic     204
Referral    181
Paid        158
REFERRAL     22
Refferal     17
referal      17
None          1
Name: count, dtype: int64

In [37]:
subscription_df["subscription_type"].isnull().sum()

1

### Observation

The `subscription_type` column contains **7 unique values (including missing values)**. The analysis reveals several inconsistencies in the representation of subscription acquisition channels.

#### Value Distribution

| Subscription Type | Count |
|-------------------|------:|
| Organic | 204 |
| Referral | 181 |
| Paid | 158 |
| REFERRAL | 22 |
| Refferal | 17 |
| referal | 17 |
| Missing | 1 |

#### Data Quality Issues Identified

The analysis identified the following issues:

1. **Case Inconsistency**
   - `Referral`
   - `REFERRAL`

2. **Spelling Inconsistency**
   - `Referral`
   - `Refferal`
   - `referal`

3. **Missing Value**
   - The column contains **1 missing value (0.17%)**.

#### Business Impact

The `subscription_type` column identifies how customers were acquired (Organic, Referral, or Paid). This information is widely used to measure marketing performance, customer acquisition efficiency, and churn rates across acquisition channels.

If these inconsistencies are not corrected, the same acquisition channel will be treated as multiple categories. For example, `Referral`, `REFERRAL`, `Refferal`, and `referal` all represent the same business concept but would appear as four separate categories in dashboards and analytical reports. This would lead to inaccurate marketing KPIs and misleading business insights.

#### Cleaning Strategy

During preprocessing:

- Convert all values to a consistent text format.
- Standardize:
  - `REFERRAL` → `Referral`
  - `Refferal` → `Referral`
  - `referal` → `Referral`
- Investigate the single missing value and handle it according to the project's predefined missing value strategy.

In [38]:
subscription_df["plan_type"].nunique(dropna=False)

3

In [39]:
subscription_df["plan_type"].value_counts(dropna=False)

plan_type
Standard    233
Premium     223
Basic       144
Name: count, dtype: int64

In [40]:
subscription_df["plan_type"].isnull().sum()

0

### Observation

The `plan_type` column contains **3 unique values** representing the subscription plans offered by the OTT platform.

#### Value Distribution

| Plan Type | Count |
|-----------|------:|
| Standard | 233 |
| Premium | 223 |
| Basic | 144 |

#### Data Quality Assessment

The analysis indicates that the `plan_type` column is of **high data quality**.

The following observations were made:

- No missing values were identified.
- No duplicate category names were found.
- No spelling inconsistencies were observed.
- No case inconsistencies (e.g., uppercase/lowercase variations) were detected.
- All plan types follow a consistent business naming convention.

#### Business Impact

The `plan_type` column is a critical business attribute used to analyze customer behavior, revenue distribution, subscription trends, and churn rates across different subscription plans.

Since the values are already standardized, the column is suitable for direct use in reporting, dashboard development, exploratory data analysis (EDA), and predictive modeling without requiring additional preprocessing.

#### Cleaning Strategy

No cleaning is required for this column. The values are already standardized and can be used directly in subsequent analysis.

In [41]:
subscription_df["contract_type"].nunique(dropna=False)

2

In [42]:
subscription_df["contract_type"].value_counts(dropna=False)

contract_type
Annual     304
Monthly    296
Name: count, dtype: int64

In [43]:
subscription_df["contract_type"].isnull().sum()

0

### Observation

The `contract_type` column contains **2 unique values** representing the billing contract duration for customer subscriptions.

#### Value Distribution

| Contract Type | Count |
|--------------|------:|
| Annual | 304 |
| Monthly | 296 |

#### Data Quality Assessment

The analysis indicates that the `contract_type` column is of **high data quality**.

The following observations were made:

- No missing values were identified.
- No duplicate category names were found.
- No spelling inconsistencies were observed.
- No case inconsistencies (e.g., uppercase/lowercase variations) were detected.
- Both contract types follow a consistent business naming convention.

#### Business Impact

The `contract_type` column is an important business attribute used to analyze customer commitment levels, renewal behavior, revenue forecasting, and churn patterns. Customers on annual contracts often exhibit different retention characteristics compared to monthly subscribers.

Since the values are already standardized, the column can be used directly for reporting, exploratory data analysis (EDA), dashboard development, and predictive modeling without requiring any preprocessing.

#### Cleaning Strategy

No cleaning is required for this column. The values are already standardized and can be used directly in subsequent analysis.

In [44]:
subscription_df["cancellation_reason"].nunique(dropna=False)

11

In [45]:
subscription_df["cancellation_reason"].value_counts(dropna=False)


cancellation_reason
None                      270
Too Expensive              86
Switched to Competitor     63
Forgot to Cancel Trial     39
No Longer Interested       34
Lack of Content            30
Payment Issues             26
Poor Streaming Quality     25
Moved to Family Plan       23
Technical Problems          3
Other                       1
Name: count, dtype: int64

In [46]:
subscription_df["cancellation_reason"].isnull().sum()

270

### Observation

The `cancellation_reason` column contains **11 unique values (including missing values)**. This column captures the reason provided by customers when cancelling their subscription.

#### Value Distribution

| Cancellation Reason | Count |
|---------------------|------:|
| Missing (NA) | 270 |
| Too Expensive | 86 |
| Switched to Competitor | 63 |
| Forgot to Cancel Trial | 39 |
| No Longer Interested | 34 |
| Lack of Content | 30 |
| Payment Issues | 26 |
| Poor Streaming Quality | 25 |
| Moved to Family Plan | 23 |
| Technical Problems | 3 |
| Other | 1 |

#### Data Quality Assessment

The categorical values are well standardized with no spelling inconsistencies, duplicate categories, or case variations.

The column contains **270 missing values (45.00%)**. However, these missing values are **expected from a business perspective**, since cancellation reasons are only applicable to customers who have cancelled their subscriptions.

Business rule validation identified only **one record** where a cancellation reason exists without a corresponding cancellation date. This represents a genuine data quality issue that should be corrected during data cleaning.

#### Business Impact

The `cancellation_reason` column provides valuable insight into why customers leave the platform and is essential for churn analysis, customer retention strategies, and business decision-making.

Since most missing values correspond to active customers, they should not be treated as missing data errors. Instead, they represent **non-applicable (N/A)** business scenarios.

Only records that violate business rules—such as a cancellation reason without a cancellation date—require corrective action.

#### Cleaning Strategy

During preprocessing:

- Preserve missing values for active customers, as they are business-valid.
- Correct the single business rule violation where a cancellation reason exists without a cancellation date.
- No standardization of category names is required, as all cancellation reasons already follow a consistent naming convention.

In [47]:
support_df["escalations"].nunique(dropna=False)

5

In [48]:
support_df["escalations"].value_counts(dropna=False)

escalations
0    328
1    243
2     60
3     15
4      2
Name: count, dtype: int64

In [49]:
support_df["escalations"].isnull().sum()

0

### Observation

The `escalations` column contains **5 unique values** representing whether a customer support ticket was escalated.

#### Value Distribution

| Escalation Value | Count |
|-----------------:|------:|
| 0 | 328 |
| 1 | 243 |
| 2 | 60 |
| 3 | 15 |
| 4 | 2 |

#### Data Quality Assessment

The column contains **no missing values**, indicating complete data availability.

However, business rule validation identified several invalid values. According to the project's business definition, the `escalations` field is a binary indicator where:

- **0 = Not Escalated**
- **1 = Escalated**

Therefore, the values **2, 3, and 4** are invalid and violate the expected domain constraints.

##### Invalid Value Summary

| Invalid Value | Count |
|--------------:|------:|
| 2 | 60 |
| 3 | 15 |
| 4 | 2 |

**Total Invalid Records:** **77 (11.88%)**

#### Business Impact

The `escalations` column is an important operational metric used to evaluate customer service performance, issue severity, and customer satisfaction.

Invalid escalation values can lead to:

- Incorrect escalation rate calculations.
- Misleading customer support KPIs.
- Inaccurate service quality reporting.
- Reduced reliability of churn prediction models, where escalation history is often an important feature.

#### Cleaning Strategy

During preprocessing:

- Validate all values against the allowed business domain (`0` and `1`).
- Replace invalid values (`2`, `3`, and `4`) with missing values (`NaN`) for further investigation.
- If business rules or source documentation become available, map the invalid values accordingly; otherwise, treat them as invalid data during analysis.

In [50]:
support_df["issue_category"].nunique(dropna=False)



18

In [51]:
support_df["issue_category"].value_counts(dropna=False)



issue_category
Critical           110
High               106
Low                106
Medium              78
Streaming           70
Content             28
Technical           26
Login               26
Billing             20
Payment             18
Account             14
Refund              12
General Inquiry     10
App                  6
Buffering            6
Subscription         6
Recommendation       4
Device               2
Name: count, dtype: int64

In [52]:
support_df["issue_category"].isnull().sum()

0

### Observation

The `issue_category` column contains **18 unique values** and has **no missing values**, indicating excellent completeness.

#### Value Distribution

| Issue Category | Count |
|----------------|------:|
| Critical | 110 |
| High | 106 |
| Low | 106 |
| Medium | 78 |
| Streaming | 70 |
| Content | 28 |
| Technical | 26 |
| Login | 26 |
| Billing | 20 |
| Payment | 18 |
| Account | 14 |
| Refund | 12 |
| General Inquiry | 10 |
| App | 6 |
| Buffering | 6 |
| Subscription | 6 |
| Recommendation | 4 |
| Device | 2 |

#### Data Quality Assessment

The `issue_category` column contains **no missing values**, and all values are consistently formatted without spelling errors, duplicate labels, or case inconsistencies.

However, the profiling analysis identified a **business schema inconsistency**. The column contains a mixture of two different business concepts:

**Issue Severity**

- Critical
- High
- Medium
- Low

**Issue Category**

- Streaming
- Content
- Technical
- Login
- Billing
- Payment
- Account
- Refund
- App
- Buffering
- Subscription
- Recommendation
- Device
- General Inquiry

Although the column has been renamed from `col_1` to `issue_category` for better readability, the source dataset combines both issue categories and severity levels within the same field.

#### Business Impact

Combining issue categories and severity levels into a single column reduces analytical flexibility and limits business reporting capabilities.

For example, it becomes difficult to answer important business questions such as:

- Which issue category generates the highest number of support tickets?
- Which issue categories receive the lowest CSAT scores?
- Which issue categories are most frequently escalated?
- How many **Critical** issues are related to **Streaming** or **Billing**?

Separating these two business dimensions would significantly improve customer support analytics, dashboard reporting, and predictive modeling.

#### Cleaning Strategy

During preprocessing:

- Retain the renamed column (`issue_category`) for improved readability.
- Identify records where the value represents **issue severity** (`Critical`, `High`, `Medium`, `Low`) instead of an issue category.
- Flag these records for business review or separate them into a dedicated `issue_severity` column if additional business rules or source documentation become available.
- Preserve valid issue categories without modification.

In [53]:
support_df["comment"].nunique(dropna=False)

89

In [54]:
support_df["comment"].value_counts(dropna=False)

comment
Poor Customer Service                   46
Billing Issue                           39
App Crash                               39
Subscription Issue                      37
Buffering                               35
Refund Requested                        35
Payment Failed                          35
None                                    32
Streaming Issue                         30
Poor Recommendation                     30
Login Problem                           25
                                        21
Problem fixed                           10
Issue resolved                          10
Great UX                                10
Everything works fine                    8
Found what I wanted                      6
Works fine on mobile                     6
Still waiting for refund                 6
Great documentary selection              6
Great quality                            6
Password reset successful                6
Account flagged incorrectly              6
Esc

In [55]:
support_df["comment"].isnull().sum()

32

### Observation

The `comment` column contains **89 unique values (including missing and blank values)** and stores free-text descriptions provided by customers regarding their support experience or reported issues.

#### Data Quality Assessment

The analysis indicates that the column primarily contains customer-written comments describing support requests, complaints, resolutions, and feedback.

Key observations include:

- **89 unique text values** were identified.
- **32 missing values (4.94%)** are present.
- **21 records contain blank strings (`""`)**, which should be treated as missing values during preprocessing.
- Most comments are meaningful and provide valuable contextual information for customer support analysis.

Examples of common comments include:

- Poor Customer Service
- Billing Issue
- Subscription Issue
- App Crash
- Buffering
- Refund Requested
- Payment Failed
- Streaming Issue
- Login Problem
- Issue Resolved
- Great UX
- Everything Works Fine

#### Data Quality Issues Identified

The following data quality issues were observed:

1. **Missing Values**
   - 32 records contain NULL values.

2. **Blank Values**
   - 21 records contain empty strings instead of meaningful text.
   - These should be standardized as missing values.

3. **Minor Text Quality Issues**
   - A few comments contain spelling mistakes (e.g., `Techncal` instead of `Technical`).
   - Since these are customer-entered free-text comments, they should generally be preserved unless text normalization is required for NLP applications.

#### Business Impact

Customer comments provide valuable qualitative information that complements structured support data.

This column can support:

- Root cause analysis
- Customer sentiment analysis
- Complaint trend analysis
- Support quality monitoring
- Natural Language Processing (NLP)
- AI-powered ticket classification

Although this column is not typically used for aggregation in dashboards, it provides significant value for advanced analytics and machine learning applications.

#### Cleaning Strategy

During preprocessing:

- Convert blank strings to NULL values.
- Remove leading and trailing whitespace.
- Preserve the original wording of customer comments.
- Correct only obvious formatting issues when necessary.
- Keep textual information unchanged to maintain data integrity for future NLP and text mining tasks.

# Numarical Analysis

In [57]:
subscription_df["monthly_charges"].describe()

count    600.000000
mean      15.927550
std       40.594656
min      -10.000000
25%       10.490000
50%       13.990000
75%       19.490000
max      999.000000
Name: monthly_charges, dtype: float64

In [56]:
subscription_df["monthly_charges"].isnull().sum()

0

In [58]:
Q1 = subscription_df["monthly_charges"].quantile(0.25)
Q3 = subscription_df["monthly_charges"].quantile(0.75)

IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print("Q1:", Q1)
print("Q3:", Q3)
print("IQR:", IQR)
print("Lower Bound:", lower_bound)
print("Upper Bound:", upper_bound)

Q1: 10.49
Q3: 19.49
IQR: 8.999999999999998
Lower Bound: -3.0099999999999962
Upper Bound: 32.989999999999995


In [59]:
subscription_df[
    (subscription_df["monthly_charges"] < lower_bound) |
    (subscription_df["monthly_charges"] > upper_bound)
][["customerid", "monthly_charges"]]

,customerid,monthly_charges
466,CUST-0467,-10.0
599,CUST-0600,999.0


### Monthly Charges

#### Observation

The `monthly_charges` column contains **600 records** with **no missing values**, indicating complete numerical coverage.

| Metric | Value |
|---|---:|
| Count | 600 |
| Mean | 15.93 |
| Median | 13.99 |
| Standard Deviation | 40.59 |
| Minimum | -10.00 |
| Q1 | 10.49 |
| Q3 | 19.49 |
| Maximum | 999.00 |
| IQR | 9.00 |

The mean (**15.93**) is higher than the median (**13.99**), indicating that extreme values are influencing the distribution.

Using the IQR method:

- Lower Bound = **-3.01**
- Upper Bound = **32.99**

Two records fall outside the statistical bounds:

| Customer ID | Monthly Charge | Classification |
|---|---:|---|
| CUST-0467 | -10.00 | Anomalous |
| CUST-0600 | 999.00 | Anomalous |

#### Data Quality Assessment

No missing values were identified.

Two anomalous values were identified:

1. **CUST-0467 → -10.00**
   - A negative subscription charge is potentially invalid under the expected business interpretation of monthly subscription revenue.

2. **CUST-0600 → 999.00**
   - This value is substantially higher than the observed distribution and exceeds the IQR upper bound of **32.99**.

These records represent **2 out of 600 records (0.33%)**.

#### Business Impact

`monthly_charges` is a key financial attribute used for:

- Revenue analysis
- Subscription plan analysis
- Customer value analysis
- Churn analysis
- Customer segmentation

The anomalous values can distort aggregate financial metrics such as average monthly revenue and may introduce noise into downstream statistical or machine learning models.

The `999.00` value is particularly influential because it is substantially higher than the typical observed charge range.

#### Cleaning Strategy

The two anomalous records should be **flagged for validation rather than automatically deleted or capped**.

Recommended actions:

- Investigate `CUST-0467` to determine whether `-10.00` is a data-entry error, refund/credit adjustment, or another valid business case.
- Investigate `CUST-0600` to determine whether `999.00` represents a legitimate charge or a data quality issue.
- Do not replace these values solely because they are statistical outliers.
- Apply a correction only after the business rule or source data confirms the appropriate value.
- Maintain an audit trail of any modification made during the cleaning process.

In [60]:
subscription_df["cltv"].describe()

count     600.000000
mean      914.875000
std       686.985279
min        60.000000
25%       383.750000
50%       625.000000
75%      1540.000000
max      2560.000000
Name: cltv, dtype: float64

In [61]:
Q1 = subscription_df["cltv"].quantile(0.25)
Q3 = subscription_df["cltv"].quantile(0.75)

IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print("Q1:", Q1)
print("Q3:", Q3)
print("IQR:", IQR)
print("Lower Bound:", lower_bound)
print("Upper Bound:", upper_bound)

Q1: 383.75
Q3: 1540.0
IQR: 1156.25
Lower Bound: -1350.625
Upper Bound: 3274.375


In [62]:
subscription_df[
    (subscription_df["cltv"] < lower_bound) |
    (subscription_df["cltv"] > upper_bound)
][["customerid", "cltv"]]

,customerid,cltv


### CLTV Analysis

**Observation:**

- Total records: **600**
- Missing values: **0**
- Mean CLTV: **914.88**
- Median CLTV: **625.00**
- Minimum: **60**
- Maximum: **2,560**
- IQR: **1,156.25**
- IQR Lower Bound: **-1,350.63**
- IQR Upper Bound: **3,274.38**
- Statistical outliers: **0**

The mean CLTV is higher than the median, indicating a right-skewed distribution. However, no statistical outliers were detected using the IQR method, and all values fall within the calculated bounds.

**Cleaning Decision:**

No cleaning is required for the `cltv` column. The original values will be retained for further analysis.

In [63]:
subscription_df["churn_score"].describe()

count    600.000000
mean      41.705000
std       19.599188
min        9.000000
25%       24.000000
50%       37.000000
75%       60.000000
max       85.000000
Name: churn_score, dtype: float64

In [64]:
Q1 = subscription_df["churn_score"].quantile(0.25)
Q3 = subscription_df["churn_score"].quantile(0.75)

IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print("Q1:", Q1)
print("Q3:", Q3)
print("IQR:", IQR)
print("Lower Bound:", lower_bound)
print("Upper Bound:", upper_bound)

Q1: 24.0
Q3: 60.0
IQR: 36.0
Lower Bound: -30.0
Upper Bound: 114.0


In [65]:
subscription_df[
    (subscription_df["churn_score"] < lower_bound) |
    (subscription_df["churn_score"] > upper_bound)
][["customerid", "churn_score"]]

,customerid,churn_score


### Churn Score Analysis

**Observation:**

- Total records: **600**
- Missing values: **0**
- Mean churn score: **41.71**
- Median churn score: **37**
- Minimum: **9**
- Maximum: **85**
- IQR: **36**
- IQR Lower Bound: **-30**
- IQR Upper Bound: **114**
- Statistical outliers: **0**

The `churn_score` column contains complete and numerically consistent data. All observed values fall within the calculated IQR boundaries, and no statistical outliers were detected.

**Cleaning Decision:**

No cleaning is required for the `churn_score` column. The original values will be retained for further analysis.

In [66]:
support_df["csat_score"].describe()

count    640.000000
mean       4.857812
std       13.060834
min       -5.000000
25%        3.000000
50%        4.000000
75%        5.000000
max      120.000000
Name: csat_score, dtype: float64

In [67]:
support_df[
    (support_df["csat_score"] < 1) |
    (support_df["csat_score"] > 5)
][["support_id", "customerid", "csat_score"]]

,support_id,customerid,csat_score
23,24,CUST-0277,0.0
24,25,CUST-0277,0.0
147,149,CUST-0277,0.0
148,150,CUST-0277,0.0
253,256,CUST-0596,-5.0
264,267,CUST-0025,-5.0
276,279,CUST-0408,120.0
296,299,CUST-0269,-5.0
309,312,CUST-0090,120.0
316,319,CUST-0354,120.0


### CSAT Score Analysis

**Observation:**

* Total records: **648**
* Non-null values: **640**
* Missing values: **8 (1.23%)**
* Mean CSAT score: **4.86**
* Median: **4**
* Minimum: **-5**
* Maximum: **120**
* Invalid CSAT records: **21**

The expected CSAT range is **1 to 5**. The analysis identified **21 invalid records** with values of **0, -5, and 120**, which fall outside the valid business range.

**Cleaning Decision:**

* Convert invalid CSAT values (`0`, `-5`, and `120`) to missing values (`NaN`).
* Preserve the existing missing values.
* Valid CSAT scores between `1` and `5` will remain unchanged.


# Create copies for cleaning

In [68]:
# Create copies for cleaning
customer_clean = customer_df.copy()
subscription_clean = subscription_df.copy()
support_clean = support_df.copy()

print("Cleaning copies created successfully.")

print("Customer :", customer_clean.shape)
print("Subscription :", subscription_clean.shape)
print("Support :", support_clean.shape)

Cleaning copies created successfully.
Customer : (600, 8)
Subscription : (600, 11)
Support : (648, 7)


In [69]:
# Convert date columns to datetime

customer_clean["dob"] = pd.to_datetime(
    customer_clean["dob"], errors="coerce"
)

subscription_clean["subscription_start_date"] = pd.to_datetime(
    subscription_clean["subscription_start_date"], errors="coerce"
)

subscription_clean["renewal_date"] = pd.to_datetime(
    subscription_clean["renewal_date"], errors="coerce"
)

subscription_clean["cancellation_date"] = pd.to_datetime(
    subscription_clean["cancellation_date"], errors="coerce"
)

support_clean["complaint_date"] = pd.to_datetime(
    support_clean["complaint_date"], errors="coerce"
)

print("Date columns converted successfully.")

Date columns converted successfully.


In [70]:
print("Customer dob:", customer_clean["dob"].dtype)
print("Subscription start:", subscription_clean["subscription_start_date"].dtype)
print("Renewal date:", subscription_clean["renewal_date"].dtype)
print("Cancellation date:", subscription_clean["cancellation_date"].dtype)
print("Complaint date:", support_clean["complaint_date"].dtype)

Customer dob: datetime64[ns]
Subscription start: datetime64[ns]
Renewal date: datetime64[ns]
Cancellation date: datetime64[ns]
Complaint date: datetime64[ns]


In [71]:
country_mapping = {
    "BD": "Bangladesh",
    "bangladesh": "Bangladesh",
    "IND": "India",
    "india": "India",
    "NP": "Nepal",
    "nepal": "Nepal"
}

customer_clean["country"] = customer_clean["country"].replace(country_mapping)

customer_clean["country"].value_counts(dropna=False)

country
Bangladesh    309
India         156
Nepal          90
None           45
Name: count, dtype: int64

In [72]:
state_mapping = {
    "DHAKA": "Dhaka",
    "Dhakaa": "Dhaka",
    "dhaka": "Dhaka",
    "Chattogram": "Chittagong"
}

customer_clean["state"] = customer_clean["state"].replace(state_mapping)

customer_clean["state"].value_counts(dropna=False)

state
Khulna         70
Sylhet         67
Dhaka          64
Rajshahi       57
Chittagong     56
None           41
West Bengal    34
Tamil Nadu     33
Maharashtra    33
Gandaki        32
Delhi          30
Karnataka      29
Lumbini        27
Bagmati        20
Kathmandu       7
Name: count, dtype: int64

In [73]:
gender_mapping = {
    "female": "Female",
    "Women": "Female",
    "F": "Female",
    "male": "Male",
    "Men": "Male",
    "M": "Male"
}

customer_clean["gender"] = customer_clean["gender"].replace(gender_mapping)

customer_clean["gender"].value_counts(dropna=False)

gender
Female     277
Male       270
None        30
Unknown     23
Name: count, dtype: int64

In [74]:
subscription_type_mapping = {
    "REFERRAL": "Referral",
    "Refferal": "Referral",
    "referal": "Referral"
}

subscription_clean["subscription_type"] = (
    subscription_clean["subscription_type"]
    .replace(subscription_type_mapping)
)

subscription_clean["subscription_type"].value_counts(dropna=False)

subscription_type
Referral    237
Organic     204
Paid        158
None          1
Name: count, dtype: int64

In [75]:
print("CUSTOMER MISSING VALUES")
print(customer_clean.isnull().sum())

print("\nSUBSCRIPTION MISSING VALUES")
print(subscription_clean.isnull().sum())

print("\nSUPPORT MISSING VALUES")
print(support_clean.isnull().sum())

CUSTOMER MISSING VALUES
customerid     0
name           0
country       45
state         41
gender        30
dob            0
interests     39
pincode       17
dtype: int64

SUBSCRIPTION MISSING VALUES
customerid                   0
subscription_start_date      0
subscription_type            1
renewal_date                 1
plan_type                    0
contract_type                0
cancellation_date          271
cancellation_reason        270
monthly_charges              0
cltv                         0
churn_score                  0
dtype: int64

SUPPORT MISSING VALUES
support_id         0
customerid         0
complaint_date     0
escalations        0
csat_score         8
issue_category     0
comment           32
dtype: int64


In [76]:
# Convert invalid CSAT scores to missing
support_clean.loc[
    (support_clean["csat_score"] < 1) |
    (support_clean["csat_score"] > 5),
    "csat_score"
] = np.nan

print("CSAT missing/invalid values:",
      support_clean["csat_score"].isnull().sum())

CSAT missing/invalid values: 29


In [77]:
# Fill missing categorical values with "Unknown"

customer_clean["country"] = customer_clean["country"].fillna("Unknown")
customer_clean["state"] = customer_clean["state"].fillna("Unknown")
customer_clean["gender"] = customer_clean["gender"].fillna("Unknown")
customer_clean["interests"] = customer_clean["interests"].fillna("Unknown")
customer_clean["pincode"] = customer_clean["pincode"].fillna("Unknown")

subscription_clean["subscription_type"] = (
    subscription_clean["subscription_type"].fillna("Unknown")
)

support_clean["comment"] = support_clean["comment"].fillna("No Comment")

print("Categorical missing values handled.")

Categorical missing values handled.


In [78]:
print("CUSTOMER")
print(customer_clean.isnull().sum())

print("\nSUBSCRIPTION")
print(subscription_clean.isnull().sum())

print("\nSUPPORT")
print(support_clean.isnull().sum())

CUSTOMER
customerid    0
name          0
country       0
state         0
gender        0
dob           0
interests     0
pincode       0
dtype: int64

SUBSCRIPTION
customerid                   0
subscription_start_date      0
subscription_type            0
renewal_date                 1
plan_type                    0
contract_type                0
cancellation_date          271
cancellation_reason        270
monthly_charges              0
cltv                         0
churn_score                  0
dtype: int64

SUPPORT
support_id         0
customerid         0
complaint_date     0
escalations        0
csat_score        29
issue_category     0
comment            0
dtype: int64


In [79]:
subscription_clean[
    subscription_clean["renewal_date"].isna()
]

,customerid,subscription_start_date,subscription_type,renewal_date,plan_type,contract_type,cancellation_date,cancellation_reason,monthly_charges,cltv,churn_score
254,CUST-0255,2021-01-29,Referral,NaT,Basic,Monthly,2021-06-14,Payment Issues,5.99,180,70


### Renewal Date

One missing `renewal_date` was identified for customer `CUST-0255`.

The available subscription information does not provide sufficient evidence to reliably reconstruct the original renewal date. Therefore, the value will be retained as missing rather than introducing an assumed date.

**Cleaning Decision:**  
Retain the missing value and avoid artificial imputation.

In [80]:
print("Customer duplicates:", customer_clean.duplicated().sum())
print("Subscription duplicates:", subscription_clean.duplicated().sum())
print("Support duplicates:", support_clean.duplicated().sum())

Customer duplicates: 0
Subscription duplicates: 0
Support duplicates: 0


In [81]:
print("Customer ID duplicates:",
      customer_clean["customerid"].duplicated().sum())

print("Subscription Customer ID duplicates:",
      subscription_clean["customerid"].duplicated().sum())

print("Support ID duplicates:",
      support_clean["support_id"].duplicated().sum())

Customer ID duplicates: 0
Subscription Customer ID duplicates: 0
Support ID duplicates: 0


In [82]:
print(customer_clean["gender"].value_counts(dropna=False))
print(customer_clean["country"].value_counts(dropna=False))
print(subscription_clean["subscription_type"].value_counts(dropna=False))
print(support_clean["csat_score"].describe())
print("\nInvalid CSAT values:",
      ((support_clean["csat_score"] < 1) |
       (support_clean["csat_score"] > 5)).sum())

gender
Female     277
Male       270
Unknown     53
Name: count, dtype: int64
country
Bangladesh    309
India         156
Nepal          90
Unknown        45
Name: count, dtype: int64
subscription_type
Referral    237
Organic     204
Paid        158
Unknown       1
Name: count, dtype: int64
count    619.000000
mean       3.544426
std        1.210478
min        1.000000
25%        3.000000
50%        4.000000
75%        5.000000
max        5.000000
Name: csat_score, dtype: float64

Invalid CSAT values: 0


In [83]:
print("Negative monthly charges:",
      (subscription_clean["monthly_charges"] < 0).sum())


print("Very high monthly charges:",
      (subscription_clean["monthly_charges"] > 100).sum())

Negative monthly charges: 1
Very high monthly charges: 1


In [84]:
print("Invalid churn scores:",
      ((subscription_clean["churn_score"] < 0) |
       (subscription_clean["churn_score"] > 100)).sum())

Invalid churn scores: 0


In [85]:
print("========== FINAL DATA QUALITY CHECK ==========")

print("\nCustomer Shape:", customer_clean.shape)
print("Subscription Shape:", subscription_clean.shape)
print("Support Shape:", support_clean.shape)

print("\nCustomer Missing:", customer_clean.isnull().sum().sum())
print("Subscription Missing:", subscription_clean.isnull().sum().sum())
print("Support Missing:", support_clean.isnull().sum().sum())

print("\nDuplicate Rows:")
print("Customer:", customer_clean.duplicated().sum())
print("Subscription:", subscription_clean.duplicated().sum())
print("Support:", support_clean.duplicated().sum())

print("\nCSAT Invalid:",
      ((support_clean["csat_score"] < 1) |
       (support_clean["csat_score"] > 5)).sum())

print("\nChurn Score Invalid:",
      ((subscription_clean["churn_score"] < 0) |
       (subscription_clean["churn_score"] > 100)).sum())

print("\nMonthly Charges Anomalies:")
print("Negative:", (subscription_clean["monthly_charges"] < 0).sum())
print("Very High (>100):",
      (subscription_clean["monthly_charges"] > 100).sum())

========== FINAL DATA QUALITY CHECK ==========

Customer Shape: (600, 8)
Subscription Shape: (600, 11)
Support Shape: (648, 7)

Customer Missing: 0
Subscription Missing: 542
Support Missing: 29

Duplicate Rows:
Customer: 0
Subscription: 0
Support: 0

CSAT Invalid: 0

Churn Score Invalid: 0

Monthly Charges Anomalies:
Negative: 1
Very High (>100): 1


## Final Data Cleaning Summary

The data cleaning and preprocessing process has been completed for the Customer, Subscription, and Support datasets.

### Completed Cleaning Activities

- Converted all identified date columns to proper `datetime` format.
- Standardized inconsistent categorical values across Customer and Subscription datasets.
- Standardized country representations.
- Standardized state representations.
- Standardized gender values.
- Standardized subscription type values.
- Converted invalid CSAT scores outside the valid 1–5 range to missing values.
- Handled categorical missing values using explicit `Unknown` labels where appropriate.
- Preserved business-meaningful missing values in cancellation fields.
- Checked duplicate rows and key-level duplicates.

### Final Validation

| Validation | Result |
|---|---:|
| Customer duplicate rows | 0 |
| Subscription duplicate rows | 0 |
| Support duplicate rows | 0 |
| Invalid CSAT scores | 0 |
| Invalid Churn Scores | 0 |
| Negative Monthly Charges | 1 |
| Very High Monthly Charges | 1 |

### Remaining Data Quality Issues

Two `monthly_charges` anomalies remain:

- `CUST-0467` → `-10.00`
- `CUST-0600` → `999.00`

These values were retained because no reliable replacement values were available from the source data. They should be flagged during downstream analysis rather than being arbitrarily modified.

One `renewal_date` value also remains missing because the correct date could not be reliably inferred from the available data.

### Conclusion

The datasets are now standardized and ready for exploratory data analysis. Remaining anomalies have been documented rather than artificially corrected, preserving data integrity and traceability.

In [87]:
import os

# Save cleaned datasets
out_dir = os.path.join("..", "data", "cleaned")
os.makedirs(out_dir, exist_ok=True)

customer_clean.to_csv(os.path.join(out_dir, "customer_clean.csv"), index=False)
subscription_clean.to_csv(os.path.join(out_dir, "subscription_clean.csv"), index=False)
support_clean.to_csv(os.path.join(out_dir, "support_clean.csv"), index=False)

print("Cleaned datasets saved successfully.")

Cleaned datasets saved successfully.
